File: src/services/accessService.js
import { BotError } from '../domain.js';

const COMMAND_LEVELS = {
  '/help': 1,
  '/ai': 1,
  '/sum': 1,
  '/rw': 1,
  '/tr': 1,
  '/g': 1,
  '/list': 1,
  '/who': 1,
  '/profile': 1,
  '/dostup': 1,
  '/tasks': 1,
  '/done': 1,
  '/ping': 1,
  '/clear': 1,
  '/task': 2,
  '/taskdel': 2,
  '/del': 2,
  '/kick': 2,
  '/add': 2,
  '/surname': 3,
  '/role': 3,
  '/setprofile': 3
};

export class AccessService {
  constructor(storage) {
    this.storage = storage;
  }

  getLevel(userId) {
    return this.storage.getAccessLevel(userId);
  }

  requireCommandAccess(userId, command) {
    const level = this.getLevel(userId);
    const needed = COMMAND_LEVELS[command] ?? 3;
    if (level < needed) {
      throw new BotError(`Недостаточно прав для команды ${command}. Требуется уровень ${needed}.`, 'FORBIDDEN');
    }
  }

  getOwnAccess(userId) {
    return this.getLevel(userId);
  }

  setAccess(actorId, targetId, level) {
    this.requireCommandAccess(actorId, '/setprofile');
    if (![1, 2, 3].includes(level)) {
      throw new BotError('Уровень доступа должен быть 1, 2 или 3.');
    }
    this.storage.ensureUser(targetId);
    this.storage.setAccessLevel(targetId, level);
    this.storage.writeAudit({ actorId, action: 'set-access-level', objectType: 'user', objectId: targetId, details: { level } });
    return level;
  }
}

File: src/main/java/com/dauren/users/AccessService.java
package com.dauren.users;

import com.dauren.services.AuditLogService;
import com.dauren.storage.SqliteStorage;

import java.util.HashMap;
import java.util.Map;

public class AccessService {
    private final SqliteStorage storage;
    private final AuditLogService audit;
    private final Map<String, Integer> commandLevels = new HashMap<>();

    public AccessService(SqliteStorage storage, AuditLogService audit) {
        this.storage = storage;
        this.audit = audit;
        init();
    }

    private void init() {
        for (String c : new String[]{"/help","/ai","/sum","/rw","/tr","/g","/list","/who","/profile","/dostup","/tasks","/done","/ping","/clear"}) commandLevels.put(c, 1);
        for (String c : new String[]{"/task","/taskdel","/del","/kick","/add"}) commandLevels.put(c, 2);
        for (String c : new String[]{"/surname","/role","/setprofile"}) commandLevels.put(c, 3);
    }

    public int levelOf(String userId) {
        return storage.getAccessLevel(userId);
    }

    public void require(String userId, String command) {
        int need = commandLevels.getOrDefault(command, 3);
        if (levelOf(userId) < need) throw new IllegalArgumentException("Недостаточно прав для " + command);
    }

    public void setAccess(String actorId, String targetId, int level) {
        require(actorId, "/setprofile");
        if (level < 1 || level > 3) throw new IllegalArgumentException("Уровень должен быть 1..3");
        storage.ensureUser(targetId, null);
        storage.setAccessLevel(targetId, level);
        audit.log(actorId, "set-access", "user", targetId, "{\"level\":" + level + "}");
    }
}
